### PR TITLE
Extract Signature Minting to a detectable feature

### PR DESCRIPTION
**Pull request by ndeto**
This PR makes Signature Minting a detectable feature. It delivers 3 signature detectable features. For ERC20/721/1155

### CODE REVIEW

File: src/core/classes/erc-721-with-quantity-signature-mintable.ts [hunk 1]
```diff
@@ -428,10 +385,16 @@ export class Erc721WithQuantitySignatureMintable implements DetectableFeature {
     } as ITokenERC721.MintRequestStructOutput;
   }
 
-  private isNFTCollection(
-    _contract: ISignatureMintERC721 | TokenERC721,
-    contractType: string,
-  ): _contract is TokenERC721 {
-    return contractType.includes("TokenERC721");
+  private async isNFTCollection() {
```
Comment (joaquim-verges) [inline src/core/classes/erc-721-with-quantity-signature-mintable.ts:388]: `isLegacyNFTContract()` to match the variable?

File: src/core/classes/erc-721-with-quantity-signature-mintable.ts [hunk 2]
```diff
@@ -428,10 +385,16 @@ export class Erc721WithQuantitySignatureMintable implements DetectableFeature {
...
+    let contractType;
+    if (hasFunction<TokenERC721>("contractType", this.contractWrapper)) {
+      contractType = ethers.utils.toUtf8String(
+        await this.contractWrapper.readContract.contractType(),
+      );
+    } else {
+      contractType = "";
```
Comment (joaquim-verges) [inline src/core/classes/erc-721-with-quantity-signature-mintable.ts:395]: just `return false`

File: src/core/classes/erc-721-with-quantity-signature-mintable.ts [hunk 3]
```diff
@@ -282,4 +362,38 @@ export class Erc721WithQuantitySignatureMinting {
...
+    } as ITokenERC721.MintRequestStructOutput;
+  }
+
+  private async isLegacyNFTContract() {
+    let contractType;
+    if (hasFunction<TokenERC721>("contractType", this.contractWrapper)) {
+      contractType = ethers.utils.toUtf8String(
+        await this.contractWrapper.readContract.contractType(),
+      );
+      return !!contractType.includes("TokenERC721");
```
Comment (joaquim-verges) [inline src/core/classes/erc-721-with-quantity-signature-mintable.ts:394]: you don't need `!!` here (it's highly discouraged) you can just do

```
      const contractType = ethers.utils.toUtf8String(
        await this.contractWrapper.readContract.contractType(),
      );
      return contractType.includes("TokenERC721");
```

File: src/core/classes/erc-721-with-quantity-signature-mintable.ts [hunk 4]
```diff
@@ -13,34 +15,41 @@ import invariant from "tiny-invariant";
...
 import { uploadOrExtractURIs } from "../../common/nft";
 import { TokensMintedWithSignatureEvent } from "contracts/SignatureDrop";
+import { DetectableFeature } from "../interfaces/DetectableFeature";
+import { FEATURE_NFT_SIGNATURE_MINTABLE } from "../../constants/erc721-features";
+import { hasFunction } from "../../common";
 
 /**
- * Enables generating dynamic ERC721 NFTs. You can authorize then some external party to mint tokens on your contract
- * @public
+ * Enables generating dynamic ERC721 NFTs with rules and an associated signature, which can then be minted by anyone securely * @public
```
Comment (joaquim-verges) [inline src/core/classes/erc-721-with-quantity-signature-mintable.ts:33]: @public should be on the next line

File: src/core/classes/erc-721-with-quantity-signature-mintable.ts [hunk 5]
```diff
@@ -199,22 +254,30 @@ export class Erc721WithQuantitySignatureMinting {
...
    *
-   * @remarks See {@link Erc721SignatureMinting.generate}
+   * @remarks See {@link Erc721WithQuantitySignatureMintable.generate}
    *
    * @param payloadsToSign - the payloads to sign
    * @returns an array of payloads and signatures
    */
   public async generateBatch(
-    payloadsToSign: PayloadToSign721withQuantity[],
+    payloadsToSign: PayloadToSign721withQuantity[] | PayloadToSign721[],
```
Comment (joaquim-verges) [inline src/core/classes/erc-721-with-quantity-signature-mintable.ts:263]: why is there `| PayloadToSign721[]` here? do we still need that type?

File: src/core/classes/erc-721-with-quantity-signature-mintable.ts [hunk 6]
```diff
@@ -199,22 +254,30 @@ export class Erc721WithQuantitySignatureMinting {
...
       await this.contractWrapper.getSignerAddress(),
     );
-    const parsedRequests: FilledSignature721WithQuantity[] = payloadsToSign.map(
-      (m) => Signature721WithQuantityInput.parse(m),
-    );
-
+    let parsedRequests;
+    if (isLegacyNFTContract) {
+      parsedRequests = payloadsToSign.map((m) =>
+        Signature721PayloadInput.parse(m),
```
[the remark this answers is not in the document]
Comment (ndeto) [reply src/core/classes/erc-721-with-quantity-signature-mintable.ts:274]: 🤦🏾‍♂️  No, we do not need them , doing a proper clean up of this now.

File: src/core/classes/erc-721-with-quantity-signature-mintable.ts [hunk 7]
```diff
@@ -269,15 +268,10 @@ export class Erc721WithQuantitySignatureMintable implements DetectableFeature {
...
-    if (isLegacyNFTContract) {
-      parsedRequests = payloadsToSign.map((m) =>
-        Signature721PayloadInput.parse(m),
-      );
-    } else {
-      parsedRequests = payloadsToSign.map((m) =>
-        Signature721WithQuantityInput.parse(m),
-      );
-    }
+    parsedRequests = payloadsToSign.map((m) =>
```
Comment (joaquim-verges) [inline src/core/classes/erc-721-with-quantity-signature-mintable.ts:271]: `const parsedRequest = ...` 🙏

**Review by joaquim-verges — APPROVED** [review]

**Pull request merged**